crates/matrix-sdk-appservice/src/lib.rs

```diff
@@ -414,19 +417,19 @@ impl AppService {
         false
     }
 
-    /// Returns a [`warp::Filter`] to be used as [`warp::serve()`] route.
-    ///
-    /// Note that if you handle any of the [application-service-specific
-    /// routes], including the legacy routes, you will break the appservice
-    /// functionality.
-    ///
-    /// Hint: [`warp::Filter`]s can be converted to an `hyper::Service` using
-    /// [`warp::service`], which allows using it with tower-compatible
-    /// frameworks such as axum.
-    ///
-    /// [application-service-specific routes]: https://spec.matrix.org/unstable/application-service-api/#legacy-routes
-    pub fn warp_filter(&self) -> warp::filters::BoxedFilter<(impl warp::Reply,)> {
-        webserver::warp_filter(self.clone())
+    /// Returns a [`Service`] that processes appservice requests.
+    pub fn service<B>(
+        &self,
+        // axum::Error is part of the signature because axum::Router::nest
+        // requires the inner service to have that exact response (body) type
+        // in 0.5.x. This will be fixed in axum 0.6.0.
+    ) -> impl HttpService<B, ResBody = UnsyncBoxBody<Bytes, axum::Error>> + Clone
```

Comment:
Unfortunately, for my usecase, I cannot manipulate this type.

Could you either give it the concrete type (which is the type returned by `webserver::router`) or make `webserver::router` public?

Reply:
Being discussed on Matrix…